src/Reflex/Backend/Socket.hs

```diff
@@ -1,192 +1,191 @@
+    -- ^ We've been asked to close, but will transmit all pending data
+    -- first (and not accept any more)
+  | Closed
+    -- ^ Hard close. Don't transmit pending data.
+
+-- | Wire a socket into the FRP network. You will likely use this to
+-- attach events to a socket that you just connected (from 'connect'),
+-- or a socket that you just accepted (from the '_aAcceptSocket' event
+-- you got when you called 'accept').
+socket
+  :: forall t m.
+     ( Reflex t
+     , PerformEvent t m
+     , PostBuild t m
+     , TriggerEvent t m
+     , MonadIO (Performable m)
+     , MonadIO m
+     )
+  => SocketConfig t
+  -> m (Socket t)
+socket (SocketConfig sock maxRx eTx eClose) = do
   (eRx, onRx) <- newTriggerEvent
   (eOpen, onOpen) <- newTriggerEvent
-  (eError, onError) <- newTriggerEvent
   (eClosed, onClosed) <- newTriggerEvent
+  (eError, onError) <- newTriggerEvent
 
-  payloadQueue <- liftIO newTQueueIO
-  closeQueue <- liftIO . atomically $ newEmptyTMVar
-  isOpenRead <- liftIO . atomically $ newEmptyTMVar
-  isOpenWrite <- liftIO . atomically $ newEmptyTMVar
-
-  ePostBuild <- getPostBuild
+  payloadQueue <- liftIO STM.newTQueueIO
+  state <- liftIO $ STM.newTVarIO Open
 
   let
-    exHandlerClose :: IOException -> IO ()
-    exHandlerClose = onError . SocketIOException
-
-    exHandlerTx :: IOException -> IO Bool
-    exHandlerTx e = do
-      mSock <- atomically . tryReadTMVar $ isOpenWrite
-      forM_ mSock $ \_ -> onError (SocketIOException e)
-      pure False
-
-    txLoop = do
-      let
-        stmTx = do
-          mSock <- tryReadTMVar isOpenWrite
-          case mSock of
-            Nothing -> pure (Left Nothing)
-            Just sock -> do
-              bs <- readTQueue payloadQueue
-              pure $ Right (sock, bs)
-        stmClose = do
-          mSock <- tryReadTMVar isOpenWrite
-          case mSock of
-            Nothing -> pure (Left Nothing)
-            Just sock -> do
-              _ <- takeTMVar closeQueue
-              pure (Left (Just sock))
-      e <- atomically $ stmClose `orElse` stmTx
-      case e of
-        Right (sock, bs) -> do
-          success <- (sendAll sock (doEncode bs) >> pure True) `catch` exHandlerTx
-          when success txLoop
-        Left (Just sock) -> do
-          void . atomically . tryTakeTMVar $ isOpenWrite
-          void . atomically . tryTakeTMVar $ isOpenRead
-          close sock `catch` exHandlerClose
-          onClosed ()
-        Left Nothing ->
-          txLoop
-
-    startTxLoop = liftIO $ do
-      mSock <- atomically $ tryReadTMVar isOpenWrite
-      forM_ mSock $ \_ -> void . forkIO $ txLoop
+    start = liftIO $ do
+      void $ forkIO txLoop
+      void $ forkIO rxLoop
+      onOpen ()
 
+      where
+        txLoop =
+          let
+            loop = do
+              mBytes <- atomically $
+                STM.readTVar state >>= \case
+                  Closed -> pure Nothing
+                  Draining -> STM.tryReadTQueue payloadQueue
```

Comment:
Does this need a `retry` like in the case for `Open`?

I'm thinking about what happens if there are multiple firings of the close event, which could cause contention on `state`, which if you were really unlucky might mean you are in the `Draining` state with things in the `payloadQueue` and `mBytes` is `Nothing` and so you close before the queue is properly drained.

I need to stare at this for longer to properly work out if this is an issue.